### PR TITLE
Finalizing the translation of chapter 7.

### PR DESCRIPTION
**Pull request by blademoon**
#75

@MKhalusova Good day Maria. This is a translation of the remainder of chapter 7. I look forward to your review.

### CODE REVIEW

File: chapters/ru/chapter7/6.mdx [hunk 1]
```diff
@@ -0,0 +1,914 @@
...
+
+Большинство документов содержит гораздо больше 128 токенов, поэтому простое обрезание входных данных до максимальной длины приведет к тому, что большая часть нашего датасета будет удалена. Вместо этого мы используем параметр `return_overflowing_tokens` для токенизации всего ввода и разбиения его на части, как мы делали в [Главе 6](/course/chapter6/4). Мы также будем использовать параметр `return_length`, чтобы автоматически возвращать длину каждого созданного фрагмента. Часто последний фрагмент будет меньше размера контекста, и мы избавимся от этих фрагментов, чтобы избежать проблем с дополнением; на самом деле они нам не нужны, поскольку у нас и так много данных.
```
Comment (MKhalusova) [inline chapters/ru/chapter7/6.mdx:178]:
```suggestion
Большинство документов содержит гораздо больше 128 токенов, поэтому простое обрезание входных данных до максимальной длины приведет к тому, что большая часть нашего датасета будет удалена. Вместо этого мы используем параметр `return_overflowing_tokens` для токенизации всего ввода и разбиения его на части, как мы делали в [Главе 6](../chapter6/4). Мы также будем использовать параметр `return_length`, чтобы автоматически возвращать длину каждого созданного фрагмента. Часто последний фрагмент будет меньше размера контекста, и мы избавимся от этих фрагментов, чтобы избежать проблем с дополнением; на самом деле они нам не нужны, поскольку у нас и так много данных.
```

File: chapters/ru/chapter7/6.mdx [hunk 2]
```diff
@@ -0,0 +1,914 @@
...
+
+В этой операции мы используем удобную особенность функции `Dataset.map()` из 🤗 Datasets, которая заключается в том, что она не требует отображения один к одному; как мы видели в [разделе 3](/course/chapter7/3), мы можем создавать батч с большим или меньшим количеством элементов, чем входной батч. Это полезно при выполнении таких операций, как аугментация или фильтрация данных, которые изменяют количество элементов. В нашем случае при токенизации каждого элемента на фрагменты заданного размера контекста мы создаем много примеров из каждого документа. Нам просто нужно убедиться, что мы удалили существующие столбцы, поскольку они имеют противоречивый размер. Если бы мы хотели их сохранить, то могли бы повторить их соответствующим образом и вернуть в рамках вызова `Dataset.map()`:
```
Comment (MKhalusova) [inline chapters/ru/chapter7/6.mdx:214]:
```suggestion
В этой операции мы используем удобную особенность функции `Dataset.map()` из 🤗 Datasets, которая заключается в том, что она не требует отображения один к одному; как мы видели в [разделе 3](../chapter7/3), мы можем создавать батч с большим или меньшим количеством элементов, чем входной батч. Это полезно при выполнении таких операций, как аугментация или фильтрация данных, которые изменяют количество элементов. В нашем случае при токенизации каждого элемента на фрагменты заданного размера контекста мы создаем много примеров из каждого документа. Нам просто нужно убедиться, что мы удалили существующие столбцы, поскольку они имеют противоречивый размер. Если бы мы хотели их сохранить, то могли бы повторить их соответствующим образом и вернуть в рамках вызова `Dataset.map()`:
```

File: chapters/ru/chapter7/6.mdx [hunk 3]
```diff
@@ -0,0 +1,914 @@
...
+Теперь, когда у нас есть готовый датасет, давайте создадим модель!
+
+<Tip>
+
+✏️ **Попробуйте!** Избавление от всех фрагментов, размер которых меньше размера контекста, не является большой проблемой, поскольку мы используем небольшие контекстные окна. При увеличении размера контекста (или если у вас корпус коротких документов) доля отбрасываемых фрагментов также будет расти. Более эффективный способ подготовки данных - объединить все токенизированные примеры в батч с маркером `eos_token_id` между ними, а затем выполнить фрагментацию на конкатенированных последовательностях. В качестве упражнения измените функцию `tokenize()`, чтобы использовать этот подход. Обратите внимание, что вам нужно установить `truncation=False` и удалить другие аргументы из токенизатора, чтобы получить полную последовательность идентификаторов токенов.
+
+</Tip>
+
+
+## Initializing a new model[[initializing-a-new-model]]
```
Comment (MKhalusova) [inline chapters/ru/chapter7/6.mdx:262]: The header needs to be translated.

File: chapters/ru/chapter7/6.mdx [hunk 4]
```diff
@@ -0,0 +1,914 @@
...
+    bos_token_id=tokenizer.bos_token_id,
+    eos_token_id=tokenizer.eos_token_id,
+)
+```
+
+С этой конфигурацией мы можем загрузить новую модель. Обратите внимание, что впервые мы не используем функцию `from_pretrained()`, поскольку фактически инициализируем модель самостоятельно:
+
+```py
+model = TFGPT2LMHeadModel(config)
+model(model.dummy_inputs)  # Builds the model
```
Comment (MKhalusova) [inline chapters/ru/chapter7/6.mdx:310]: Minor nit: you might want to translate the comment.

File: chapters/ru/chapter7/6.mdx [hunk 5]
```diff
@@ -0,0 +1,914 @@
...
+        losses.append(accelerator.gather(outputs.loss))
+    loss = torch.mean(torch.cat(losses))
+    try:
+        perplexity = torch.exp(loss)
+    except OverflowError:
+        perplexity = float("inf")
+    return loss.item(), perplexity.item()
+```
+
+С помощью функции `evaluate()` мы можем сообщать о потерях и [перплексии](/course/chapter7/3) через регулярные промежутки времени. Далее мы переопределим нашу модель, чтобы убедиться, что мы снова обучаемся с нуля:
```
Comment (MKhalusova) [inline chapters/ru/chapter7/6.mdx:772]:
```suggestion
С помощью функции `evaluate()` мы можем сообщать о потерях и [перплексии](../chapter7/3) через регулярные промежутки времени. Далее мы переопределим нашу модель, чтобы убедиться, что мы снова обучаемся с нуля:
```

File: chapters/ru/chapter7/6.mdx [hunk 6]
```diff
@@ -0,0 +1,914 @@
...
+accelerator = Accelerator(fp16=True)
+
+model, optimizer, train_dataloader, eval_dataloader = accelerator.prepare(
+    model, optimizer, train_dataloader, eval_dataloader
+)
+```
+
+<Tip>
+
+🚨 Если вы проводите обучение на TPU, вам нужно будет перенести весь код, начиная с ячейки выше, в выделенную функцию обучения. Подробнее смотрите [Главу 3](/course/chapter3).
```
Comment (MKhalusova) [inline chapters/ru/chapter7/6.mdx:800]:
```suggestion
🚨 Если вы проводите обучение на TPU, вам нужно будет перенести весь код, начиная с ячейки выше, в выделенную функцию обучения. Подробнее смотрите [Главу 3](../chapter3/1).
```

File: chapters/ru/chapter7/7.mdx [hunk 7]
```diff
@@ -0,0 +1,1203 @@
...
+
+<Tip>
+
+💡 Модели, основанные только на энкодере, такие как BERT, как правило, отлично справляются с извлечением ответов на фактоидные вопросы типа "Кто изобрел архитектуру трансформера?", но плохо справляются с открытыми вопросами типа "Почему небо голубое?". В таких сложных случаях для синтеза информации обычно используются модели энкодеров-декодеров, такие как T5 и BART, что очень похоже на [сумризацию текста](/course/chapter7/5). Если вам интересен этот тип *генеративных* ответов на вопросы, рекомендуем ознакомиться с нашим [демо](https://yjernite.github.io/lfqa.html) основанным на [датасете ELI5(https://huggingface.co/datasets/eli5).
```
Comment (MKhalusova) [inline chapters/ru/chapter7/7.mdx:37]:
```suggestion
💡 Модели, основанные только на энкодере, такие как BERT, как правило, отлично справляются с извлечением ответов на фактоидные вопросы типа "Кто изобрел архитектуру трансформера?", но плохо справляются с открытыми вопросами типа "Почему небо голубое?". В таких сложных случаях для синтеза информации обычно используются модели энкодеров-декодеров, такие как T5 и BART, что очень похоже на [сумризацию текста](/course/chapter7/5). Если вам интересен этот тип *генеративных* ответов на вопросы, рекомендуем ознакомиться с нашим [демо](https://yjernite.github.io/lfqa.html) основанным на [датасете ELI5](https://huggingface.co/datasets/eli5).
```

File: chapters/ru/chapter7/7.mdx [hunk 8]
```diff
@@ -0,0 +1,1203 @@
...
+```
+
+В качестве меток будут использоваться индексы токенов, начинающих и заканчивающих ответ, а задача модели - предсказать один начальный и конечный логит для каждого токена на входе, при этом теоретические метки будут выглядеть следующим образом:
+
+<div class="flex justify-center">
+<img class="block dark:hidden" src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter7/qa_labels.svg" alt="One-hot encoded labels for question answering."/>
+<img class="hidden dark:block" src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter7/qa_labels-dark.svg" alt="One-hot encoded labels for question answering."/>
+</div>
+
+В данном случае контекст не слишком длинный, но некоторые примеры в датасете имеют очень длинные контексты, которые превысят установленную нами максимальную длину (которая в данном случае равна 384). Как мы видели в [Главе 6] (/course/chapter6/4), когда изучали внутреннее устройство конвейера `question-answering`, мы будем работать с длинными контекстами, создавая несколько обучающих признаков из одной выборки нашего датасета, со скользящим окном между ними.
```
Comment (MKhalusova) [inline chapters/ru/chapter7/7.mdx:188]:
```suggestion
В данном случае контекст не слишком длинный, но некоторые примеры в датасете имеют очень длинные контексты, которые превысят установленную нами максимальную длину (которая в данном случае равна 384). Как мы видели в [Главе 6](../chapter6/4), когда изучали внутреннее устройство конвейера `question-answering`, мы будем работать с длинными контекстами, создавая несколько обучающих признаков из одной выборки нашего датасета, со скользящим окном между ними.
```

File: chapters/ru/chapter7/7.mdx [hunk 9]
```diff
@@ -0,0 +1,1203 @@
...
+```
+
+Как мы можем видеть, наш пример был разбит на четыре входа, каждый из которых содержит вопрос и часть контекста. Обратите внимание, что ответ на вопрос ("Bernadette Soubirous") появляется только в третьем и последнем входе, поэтому, работая с длинными контекстами таким образом, мы создадим несколько обучающих примеров, в которых ответ не будет включен в контекст. Для этих примеров метками будут `start_position = end_position = 0` (таким образом мы предсказываем токен `[CLS]`). Мы также зададим эти метки в неудачном случае, когда ответ был усечен, так что у нас есть только его начало (или конец). Для примеров, где ответ полностью находится в контексте, метками будут индекс токена, с которого начинается ответ, и индекс токена, на котором ответ заканчивается.
+
+Датасет предоставляет нам начальный символ ответа в контексте, а прибавив к нему длину ответа, мы можем найти конечный символ в контексте. Чтобы сопоставить их с индексами токенов, нам нужно использовать сопоставление смещений, которое мы изучали в [Главе 6](/course/chapter6/4). Мы можем настроить наш токенизатор на их возврат, передав `return_offsets_mapping=True`:
```
Comment (MKhalusova) [inline chapters/ru/chapter7/7.mdx:220]:
```suggestion
Датасет предоставляет нам начальный символ ответа в контексте, а прибавив к нему длину ответа, мы можем найти конечный символ в контексте. Чтобы сопоставить их с индексами токенов, нам нужно использовать сопоставление смещений, которое мы изучали в [Главе 6](../chapter6/4). Мы можем настроить наш токенизатор на их возврат, передав `return_offsets_mapping=True`:
```

File: chapters/ru/chapter7/7.mdx [hunk 10]
```diff
@@ -0,0 +1,1203 @@
...
+
+<Youtube id="BNy08iIWVJM"/>
+
+{:else}
+
+<Youtube id="VN67ZpN33Ss"/>
+
+{/if}
+
+Модель выведет логиты для начальной и конечной позиций ответа во входных идентификаторах, как мы видели при изучении [конвейера `question-answering`](/course/chapter6/3b). Шаг постобработки будет аналогичен тому, что мы делали там, так что вот краткое напоминание о том, что мы делали:
```
Comment (MKhalusova) [inline chapters/ru/chapter7/7.mdx:532]:
```suggestion
Модель выведет логиты для начальной и конечной позиций ответа во входных идентификаторах, как мы видели при изучении [конвейера `question-answering`](../chapter6/3b). Шаг постобработки будет аналогичен тому, что мы делали там, так что вот краткое напоминание о том, что мы делали:
```

File: chapters/ru/chapter7/7.mdx [hunk 11]
```diff
@@ -0,0 +1,1203 @@
...
+
+</Tip>
+
+{#if fw === 'pt'}
+
+Если вы хотите более глубоко погрузиться в тренировочный цикл, мы покажем вам, как сделать то же самое с помощью 🤗 Accelerate.
+
+## Пользовательский цикл обучения[[a-custom-training-loop]]
+
+Теперь давайте посмотрим на полный цикл обучения, чтобы вы могли легко настроить нужные вам части. Он будет очень похож на цикл обучения в [Главе 3](/course/chapter3/4), за исключением цикла оценки. Мы сможем регулярно оценивать модель, поскольку больше не ограничены классом `Trainer`.
```
Comment (MKhalusova) [inline chapters/ru/chapter7/7.mdx:1011]:
```suggestion
Теперь давайте посмотрим на полный цикл обучения, чтобы вы могли легко настроить нужные вам части. Он будет очень похож на цикл обучения в [Главе 3](../chapter3/4), за исключением цикла оценки. Мы сможем регулярно оценивать модель, поскольку больше не ограничены классом `Trainer`.
```

File: chapters/ru/chapter7/7.mdx [hunk 12]
```diff
@@ -0,0 +1,1203 @@
...
+{/if}
+
+## Использование дообученной модели[[using-the-fine-tuned-model]]
+
+Мы уже показали вам, как можно использовать модель, дообучение которой мы проводили на Model Hub, с помощью виджета инференса. Чтобы использовать ее локально в `pipeline`, нужно просто указать идентификатор модели:
+
+```py
+from transformers import pipeline
+
+# Replace this with your own checkpoint
```
Comment (MKhalusova) [inline chapters/ru/chapter7/7.mdx:1184]: Please translate the comment

File: chapters/ru/chapter7/8.mdx [hunk 13]
```diff
@@ -0,0 +1,22 @@
+# Освоение NLP[[mastering-nlp]]
+
+<CourseFloatingBanner
+    chapter={7}
+    classNames="absolute z-10 right-0 top-0"
+/>
+
+Если вы дошли до конца курса, поздравляем - теперь у вас есть все знания и инструменты, необходимые для решения (почти) любой задачи НЛП с помощью 🤗 Transformers и экосистемы Hugging Face!
```
Comment (MKhalusova) [inline chapters/ru/chapter7/8.mdx:8]:
```suggestion
Если вы дошли до конца курса, поздравляем - теперь у вас есть все знания и инструменты, необходимые для решения (почти) любой задачи NLP с помощью 🤗 Transformers и экосистемы Hugging Face!
```

Comment (MKhalusova) [reply chapters/ru/chapter7/8.mdx:8]: I don't think NLP should be translated. And if you insist on translating, then I don't think it's a correct abbreviation (I think it's "Обработка естественного языка", but I'm not sure).

File: chapters/ru/chapter7/7.mdx (hunk 12 above)
Comment (blademoon) [reply chapters/ru/chapter7/7.mdx]: Fix in 5bfa31b66758dfd355e98081e97c35dcffa324c7

File: chapters/ru/chapter7/6.mdx [hunk 14]
```diff
@@ -0,0 +1,914 @@
...
+
+Код на Python в изобилии доступен в таких репозиториях кода, как GitHub, и мы можем использовать его для создания датасета путем поиска каждого розитория Python. Именно такой подход был использован в книге [Transformers textbook](https://learning.oreilly.com/library/view/natural-language-processing/9781098136789/) для предварительного обучения большой модели GPT-2. Используя дамп GitHub объемом около 180 ГБ, содержащий примерно 20 миллионов файлов Python под названием `codeparrot`, авторы создали датасет, которым затем поделились на [Hugging Face Hub](https://huggingface.co/datasets/transformersbook/codeparrot).
+
+Однако обучение на полном корпусе требует много времени и вычислений, а нам нужно только подмножество датасетов, связанных со стеком data science на Python. Итак, давайте начнем с фильтрации датасета `codeparrot` по всем файлам, включающим любую из библиотек из этого стека. Из-за большого размера датасета мы хотим избежать его загрузки; вместо этого мы будем использовать функцию потоковой передачи (streaming), чтобы фильтровать его на лету. To help us filter the code samples using the libraries we mentioned earlier, we'll use the following function:
```
[the remark this answers is not in the document]
Comment (blademoon) [reply chapters/ru/chapter7/6.mdx]: Fix in 9af60804271ee2bdb4469f38fa038793d4dfd343

File: chapters/ru/chapter7/6.mdx (hunk 3 above)
Comment (blademoon) [reply chapters/ru/chapter7/6.mdx]: Fix in 844825dd78d6cca38b65432722f6138295dab7f3

File: chapters/ru/chapter7/6.mdx (hunk 4 above)
Comment (blademoon) [reply chapters/ru/chapter7/6.mdx]: Fix in 844825dd78d6cca38b65432722f6138295dab7f3
